ansys/mapdl/core/logging.py

```diff
@@ -0,0 +1,499 @@
+FILE_MSG_FORMAT = STDOUT_MSG_FORMAT
+
+DEFAULT_STDOUT_HEADER = """
+| Level    | Instance        | Module           | Function             | Message
+|----------|-----------------|------------------|----------------------|--------------------------------------------------------
+"""
+DEFAULT_FILE_HEADER = DEFAULT_STDOUT_HEADER
+
+NEW_SESSION_HEADER = f"""\n=====================================================================================================================================
+========================                      NEW SESSION - {datetime.now().strftime("%m/%d/%Y, %H:%M:%S")}                             ========================
+====================================================================================================================================="""
+
+string_to_loglevel = {
+    'DEBUG': DEBUG,
+    'INFO': INFO,
+    'WARN': WARN,
+    'ERROR': ERROR,
+    'CRITICAL': CRITICAL,
+}
+
+
+## Code
+
+class PyansysCustomAdapter(logging.LoggerAdapter):
+    """
+    This is key to keep the reference to the MAPDL instance name dynamic.
+    If we use the standard approach which is supplying ``extra`` input to the logger, we
+    would need to keep inputting MAPDL instances every time we do a log.
+
+    Using adapters we just need to especify the MAPDL instance we refer to once.
+    """
+
+    level = None  # This is maintained for compatibility with ``supress_logging``, but it does nothing.
+    _file_handler = None
+    _stdout_handler = None
+
+    def __init__(self, logger, extra=None):
+        self.logger = logger
+        self.extra = extra
+        self._file_handler = logger._file_handler
+        self._std_out_handler = logger._std_out_handler
+
+    def process(self, msg, kwargs):
+        kwargs['extra'] = {}
+        # This are the extra parameters sent to log
+        kwargs['extra']['instance_name'] = self.extra['name'] # here self.extra is the argument pass to the ``PyAnsys
+        return msg, kwargs
+
+    def log_to_file(self, filename=FILE_NAME, level=LOG_LEVEL):
+        """
+        Add file handler to logger.
+
+        Parameters
+        ----------
+        filename : str, optional
+            Name of the file where the logs are recorded. By default FILE_NAME
+        level : str, optional
+            Level of logging. E.x. 'DEBUG'. By default LOG_LEVEL
+        """
+
+        self.logger = add_file_handler(self.logger, filename=filename, level=level, write_headers=True)
+        self._file_handler = self.logger._file_handler
+
+    def log_to_stdout(self, level=LOG_LEVEL):
+        """
+        Add standard output handler to the logger.
+
+        Parameters
+        ----------
+        level : str, optional
+            Level of logging record. By default LOG_LEVEL
+        """
+
+        self.logger = add_stdout_handler(self.logger, level=level)
+        self._std_out_handler = self.logger._std_out_handler
```

Comment:
Add in a check to verify if we've already a stdout handler and raise an exception if the user tries to add it twice.

Reply:
Done!